Add initial support for powerpc64le initialization.

BODAPATIMAHESH opened the pull request:
Added header definitions for PPC64LE.

Implemented support to construct and report processor topology information including processor, core, cluster, package, PVR, and cache levels (L1i, L1d, L2, and L3).

Test:

Built and executed cpu_info on a PPC64LE Linux machine. Confirmed that it accurately reports logical processors, cores, clusters, packages, and cache information.

Ran unit tests and verified that there were no regressions.

BODAPATIMAHESH committed:
Add initial support for powerpc64le initialization.

    * Adds header definitions for PPC64le
    * Adds support to construct the processor, core, cluster, package
      and cache(L1i,L1d,L2 and L3) information reported by the system.

    Test: Build and ran cpu_info on PPC64le linux machine. confirmed
          that it properly reports the logical processors, cores, clusters,
          packages and cache information.

BODAPATIMAHESH commented on test/name/power-features.cc:
```diff
@@ -0,0 +1,41 @@
+		int b = (uint32_t) getauxval(AT_HWCAP2);
+	#endif  //CPUINFO_MOCK //
+
+	#if (b & CPUINFO_POWERPC_LINUX_FEATURE_ARCH_3_1)
```
@malfet when AT_hwcap2 supports the feature arch_3_1(Power10) then cpuinfo_initialization should set these features mma,vsx 
        _```
#if (b & CPUINFO_POWERPC_LINUX_FEATURE_ARCH_3_1)
                EXPECT_EQ(0, cpuinfo_has_powerpc_htm());
                EXPECT_EQ(1, cpuinfo_has_powerpc_mma());
                EXPECT_EQ(1, cpuinfo_has_powerpc_vsx());
```_

malfet replied:
`#if` is a [preprocessor](https://en.wikipedia.org/wiki/C_preprocessor) directive, isn't it?
And `b` is not defined when preprocessor is invoked, so all those conditions will simply never be compiled, aren't they?

So, I would like to see some sort of a thorough test plan for this PR before accepting it

BODAPATIMAHESH commented on src/linux/processors.c:
```diff
@@ -280,6 +283,31 @@ bool cpuinfo_linux_get_processor_package_id(uint32_t processor, uint32_t package
+			*online_status_ptr = online_status;
+			return true;
+	} else {
+	cpuinfo_log_info(
+	"failed to parse online status for processor %" PRIu32 " from %s", processor, processor_online_filename);
+	return false;
```
yeah, I will fix all the indent errors

BODAPATIMAHESH replied:
@malfet I have fixed indentation. could you review it

Commit:
Fixing Indentation

BODAPATIMAHESH commented:
@malfet could you review the patch again

BODAPATIMAHESH commented:
Hi @malfet , Did you get a chance to review this patch?

BODAPATIMAHESH commented:
@malfet I am waiting for your review.

BODAPATIMAHESH commented:
@fbarchard @malfet can you review this.

malfet commented on test/name/power-features.cc:
```diff
@@ -0,0 +1,41 @@
+	int b = (uint32_t)getauxval(AT_HWCAP2);
+#endif // CPUINFO_MOCK //
+
+#if (b & CPUINFO_POWERPC_LINUX_FEATURE_ARCH_3_1)
```
Can you please explain to me, when something like that would be evaluated to true?

BODAPATIMAHESH replied:
Yes , you are right . I have updated the code.

malfet reviewed (changes requested):
I'm still confused about use of `#if (b & FOOBAR)` conditions in the test

BODAPATIMAHESH committed:
Replacing the preprocessor conditions with if-elseif, as the
preprocessor code will never be compiled.

BODAPATIMAHESH committed:
Fixed the indentation and clang-format

BODAPATIMAHESH commented:
I have updated the code. Please review it @malfet

BODAPATIMAHESH commented:
@malfet can you review the latest code.

malfet commented on include/cpuinfo.h:
```diff
@@ -671,9 +686,14 @@ struct cpuinfo_core {
 	/** Clock rate (non-Turbo) of the core, in Hz */
 	uint64_t frequency;
+
+	bool disabled;
```
Please mention in PR description that it adds a new property (though I though there were already one) And where is it initialized for other architectures?

BODAPATIMAHESH replied:
Hi @malfet

BODAPATIMAHESH commented on include/cpuinfo.h:
```diff
@@ -671,9 +686,14 @@ struct cpuinfo_core {
 #elif CPUINFO_ARCH_ARM || CPUINFO_ARCH_ARM64
 	/** Value of Main ID Register (MIDR) for this core */
```
MIDR for ARM and ARM64

BODAPATIMAHESH committed:
Remove redundant 'disabled' declaration for PowerPC64 Linux

The 'disabled' field was declared for configurations other than SMT8
mode, but it is already handled during PowerPC64 Linux system
initialization. This change removes the redundant code.
make test:
    Start 1: init-test
1/3 Test #1: init-test ........................   Passed    0.02 sec
    Start 2: get-current-test
2/3 Test #2: get-current-test .................   Passed    0.02 sec
    Start 3: power-features-test
3/3 Test #3: power-features-test ..............   Passed    0.02 sec

100% tests passed, 0 tests failed out of 3

malfet reviewed (changes requested):
Please remove `cpuinfo_linux_get_processor_online_status` and then it looks like a reasonable structured PR

BODAPATIMAHESH committed:
PPC64le:This patch introduces the following updates.

Fixes incorrect processor linux_ids when running in modes other than
SMT8.
Corrects cache count reporting when SMT is disabled.
Removed the cpuinfo_linux_get_processor_online_status and added it in
the powerpc specific code and optimized the usage of it.
Cleans up the codebase to improve readability and maintainability.

anjiltech reviewed (approved):
Updated changes looks fine. Will let @malfet suggest next actions/approval.

BODAPATIMAHESH committed:
Merge branch 'main' into main_ppc64le

BODAPATIMAHESH commented:
@malfet could you re-review the PR
